Python: merge expressions of interpreters

FRidh opened the pull request:
- Merge CPython expressions
- Merge PyPy expressions
- Improve cross-compilation of Python packages
- Drop `python.majorVersion` attribute
- Introduce derivations for prebuilt PyPy binaries.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FRidh committed:
CPython: merge expressions of interpreters

Each time a new major/minor version of CPython was released, a new
expression would be written, typically copied from the previous release.
Often fixes are only made in the current/latest release. By merging the
expressions it's more likely that modifications end up in all versions,
as is likely intended.

This commit introduces one expression for Python 3, and another for 2.7.
These two may also be merged, but it will result in a lot of extra
conditionals making the expression harder to follow.

A common passthru is introduced for CPython and PyPy.

python 2.7: use common passthru

FRidh commented:
cc @LnL7 for Darwin
cc @Mic92 for cross

FRidh committed:
pypy: merge 2.7 and 3.5 into a single expression

This commit merges the two expressions in a single one, using
the passthru function that is shared with CPython.

Mic92 commented:
I will have a look at cross-compiling once the conflict is resolved.

FRidh commented:
@Mic92 not sure whether to trust this function:
```
$(nix-build -A pkgsCross.musl64.python3)/bin/python -c "import platform; print(platform.python_compiler())"
GCC 7.4.0
```

FRidh committed:
Python: improve cross-compilation

This changeset allows for cross-compilation of Python packages. Packages
built with buildPythonPackage are not allowed to refer to the build
machine. Executables that have shebangs will refer to the host.

Mic92 commented:
@FRidh what's wrong with the output? Afaik we build with gcc for musl.

Mic92 commented on pkgs/development/interpreters/python/build-python-package-setuptools.nix:
```diff
@@ -26,13 +26,13 @@ in attrs // {
     runHook preBuild
     cp ${setuppy} nix_run_setup
-    ${python.interpreter} nix_run_setup ${lib.optionalString (setupPyBuildFlags != []) ("build_ext " + (lib.concatStringsSep " " setupPyBuildFlags))} bdist_wheel
+    ${python.crossPython.interpreter} nix_run_setup ${lib.optionalString (setupPyBuildFlags != []) ("build_ext " + (lib.concatStringsSep " " setupPyBuildFlags))} bdist_wheel
```
cc @Ericson2314 might have an opinion on this.

FRidh replied:
Not sure how to consider the sliding window here. We need to have a reference to the `buildPackages` version, otherwise it gets very messy.

Ericson2314 replied:
Yes good catch and good fix. `pythonForBuild` is a perfect name such that I know the intent is correct :).

FRidh commented:
yes, I mixed host and build up when I wrote that.

Mic92 commented on pkgs/development/python-modules/bootstrapped-pip/default.nix:
```diff
@@ -42,7 +42,7 @@ in stdenv.mkDerivation rec {
 
     # install pip binary
-    echo '#!${python.interpreter}' > $out/bin/pip
+    echo '#!${python.pythonForBuild.interpreter}' > $out/bin/pip
```
Is this correct? Would this not put the wrong python into pip?

Ericson2314 replied:
Yeah this looks like an overbite. Maybe a this a holdover from `crossPython` being renamed?

Ericson2314 commented on pkgs/development/interpreters/python/build-python-package-setuptools.nix:
```diff
@@ -47,8 +47,8 @@ in attrs // {
-      mkdir -p $tmp_path/${python.sitePackages}
+      export PYTHONPATH="$tmp_path/${python.pythonForBuild.sitePackages}:$PYTHONPATH"
+      mkdir -p $tmp_path/${python.pythonForBuild.sitePackages}
       ${bootstrapped-pip}/bin/pip install -e . --prefix $tmp_path >&2
```
Probably should be `python.pythonForBuild.bootstrapped-pip`. This will probably allow for https://github.com/NixOS/nixpkgs/pull/53123#discussion_r244846036 to be fixed.

Ericson2314 commented:
Very glad this is happening! I started something like the once, but it appears I lost it haha.

FRidh committed:
Python: drop `python.majorVersion`

Drop `python.majorVersion`. For Python language version, use `python.pythonVersion`.
For implementation version, use `python.sourceVersion`.

Some expressions were broken. Those that were identified were fixed.

fixup major

FRidh committed:
pythonInterpreters.pypy{27,35}_prebuilt: init at 6.0.0

These interpreters are prebuilt by upstream and patched using patchelf.
They are primarily added for testing purposes and development on the
non-prebuilt PyPy interpreters as it can speed up translation
significantly.

Ericson2314 commented on pkgs/development/interpreters/python/build-python-package-common.nix:
```diff
@@ -10,7 +9,7 @@
 
 attrs // {
-  buildInputs = buildInputs ++ [ bootstrapped-pip ];
+  buildInputs = buildInputs ++ [ python.pythonForBuild.pkgs.bootstrapped-pip ];
```
Bootstrapped pip is used at build time and so should be a `nativeBuildInput`.

Ericson2314 replied:
Alternatively, since you use it explicitly below, maybe it doesn't need to be any sort of imput.

FRidh replied:
Good catch. I don't think we actually need it as a build input. We call it already from the store path, and Python packages should never call or import from pip. Doesn't mean they won't do it though...

Ericson2314 commented on pkgs/development/interpreters/python/build-python-package-common.nix:
```diff
@@ -24,7 +23,7 @@ attrs // {
 
     pushd dist
-    ${bootstrapped-pip}/bin/pip install *.whl --no-index --prefix=$out --no-cache ${toString installFlags} --build tmpbuild
+    ${python.pythonForBuild.pkgs.bootstrapped-pip}/bin/pip install *.whl --no-index --prefix=$out --no-cache ${toString installFlags} --build tmpbuild
```
N.B. this is fine, but won't need it once `bootstrapped-pip` is a `nativeBuildInput`.

Ericson2314 reviewed (commented):
It's looking really good!

Mic92 commented on pkgs/development/interpreters/python/pypy/prebuilt.nix:
```diff
@@ -0,0 +1,105 @@
+    pushd $out
+    ls -l .
+    find {lib,lib_pypy*} -name "*.so" -exec patchelf --replace-needed "libbz2.so.1.0" "libbz2.so.1" {} \;
+    find {lib,lib_pypy*} -name "*.so" -exec patchelf --set-rpath ${stdenv.lib.makeLibraryPath deps} {} \;
```
In future `autoPatchelfHook` could make this process easier and less error-prone.
